src/librustc_middle/ty/relate.rs

```diff
@@ -559,7 +560,27 @@ pub fn super_relate_consts<R: TypeRelation<'tcx>>(
                     }
                 }
 
-                // FIXME(const_generics): handle `ConstValue::ByRef`.
+                (ConstValue::ByRef { .. }, ConstValue::ByRef { .. }) => {
+                    // `a` and `b` are already fully resolved, so we can
+                    // use `reveal_all` here.
+                    let param_env = ty::ParamEnv::reveal_all();
+                    let a_destructured = tcx.destructure_const(param_env.and(a));
+                    let b_destructured = tcx.destructure_const(param_env.and(b));
+
+                    // Both the variant and each field have to be equal.
+                    if a_destructured.variant == b_destructured.variant {
+                        for (a_field, b_field) in
+                            a_destructured.fields.iter().zip(b_destructured.fields.iter())
+                        {
+                            relation.consts(a_field, b_field)?;
```

Comment:
@oli-obk @matthewjasper What do you think of this use of `destructure_const`?

Reply:
`destructure_const` panics if the type isn't an array, adt or tuple, so it looks incorrect to not have a check on the type.

Reply:
it is. breaks with https://github.com/rust-lang/rust/issues/63322

currently working on this

Reply:
It should allow closures as well, IMO, although I'm not sure closures should be considered structurally matcheable...

Reply:
Should be fixed, don't know if there are other types which have `ConstValue::ByRef` and should work rn.

Reply:
Oh yeah you can't have closures in a `const` type, so you can mark closures as not structurally matcheable.

Reply:
Wait, hang on, it can't be `ty::Closure`, maybe you were still talking about `dyn Trait`.

Reply:
both examples in #63322 end up relying on `&dyn Trait` afaict